Allow gecko to provide the already-applied sticky offset to WR

staktrace opened the pull request:
Fixes the remaining issue in #1912.

r? @mrobinson

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/1977)
<!-- Reviewable:end -->

mrobinson commented on webrender/src/clip_scroll_node.rs:
```diff
@@ -454,6 +457,14 @@ impl ClipScrollNode {
         }
     }
 
+    fn clamp_to_bounds(
```
It probably makes more sense for this to be a trait on f32 if possible. The self value is unused here.

mrobinson commented on webrender/src/clip_scroll_node.rs:
```diff
@@ -478,54 +489,73 @@ impl ClipScrollNode {
+            debug_assert!(sticky_offset.y + info.applied_offset.y >= 0.0);
         }
 
         if sticky_offset.y == 0.0 {
```
Hrm. We don't need to test sticky_offset.y plus the applied offset?

staktrace replied:
You're right, we do. When I thought it through before I thought this was correct but thinking it through again now you're right. (For posterity, the specific thought-example is the case where gecko paints something with a top-offset already applied, and then with the async scroll we completely unapply that offset and apply a bottom-offset).

mrobinson commented on wrench/src/yaml_frame_writer.rs:
```diff
@@ -999,6 +999,9 @@ impl YamlFrameWriter {
                     yaml_node(&mut v, "horizontal-offset-bounds", Yaml::Array(horizontal));
                     yaml_node(&mut v, "vertical-offset-bounds", Yaml::Array(vertical));
+
+                    f32_node(&mut v, "applied-offset-x", item.applied_offset.x);
```
Thanks for adding wrench support! I think it would be good to add a test for this as well, since there is a risk of breaking such a particular use case.

staktrace replied:
I'm not entirely sure how to do this but I'll ask around.

mrobinson commented on webrender/src/clip_scroll_node.rs:
```diff
@@ -478,54 +489,73 @@ impl ClipScrollNode {
+        // the calling code, stored in info.applied_offset, and the extra amount we
+        // computed as a result of scrolling, stored in sticky_offset) needs to be
+        // clamped to the provided bounds.
+        sticky_offset.y = self.clamp_to_bounds(sticky_offset.y + info.applied_offset.y, &info.vertical_offset_bounds) - info.applied_offset.y;
```
These lines extend beyond 100 characters.

mrobinson commented on webrender_api/src/display_item.rs:
```diff
@@ -155,6 +155,13 @@ pub struct StickyFrameDisplayItem {
+    /// has been scrolled down, such that the sticky item's position needed to be offset
+    /// downwards by `applied_offset.y`. A negative y component corresponds to the upward
+    /// offset applied due to bottom-stickiness. The x-axis works analogously.
+    pub applied_offset: LayoutVector2D,
```
Maybe this should be called something like `previously_applied_offset` to avoid confusing it with the current offset provided by WebRender?

staktrace replied:
I was thinking of that too but wasn't sure how about variable name length. I can change it.

staktrace commented:
Updated. For the clamp function I just made a little lambda inside the main function which I think is cleaner than before. While writing the reftests I also discovered that neither `sticky_offset.y == 0` nor `sticky_offset.y + info.applied_offset.y == 0` were correct, so I fixed that up better and added a comment to indicate what happened there. Hopefully this version addresses all of your comments - let me know if there's anything else you'd like to see changed!

staktrace commented:
Travis failures are just timeouts; please ask bors to retry if the updated patch looks good to you.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #1981) made this pull request unmergeable. Please resolve the merge conflicts.

staktrace commented:
Rebased ^. Can somebody with permissions stamp this with r=mrobinson please? Thanks!

glennw commented:
@bors-servo r=mrobinson

Thanks!

bors-servo commented:
:pushpin: Commit 89af6d3 has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson 89af6d3833195fbc9c4e7a3ead985de328bbf2ae -->

bors-servo commented:
:hourglass: Testing commit 89af6d3833195fbc9c4e7a3ead985de328bbf2ae with merge ec472226b65d279957c3aebbe211d1c3bbc20ca5...

bors-servo commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/servo/webrender/build/1.0.6234), [status-travis](https://travis-ci.org/servo/webrender/builds/297772085?utm_source=github_status&utm_medium=notification)
Approved by: mrobinson
Pushing ec472226b65d279957c3aebbe211d1c3bbc20ca5 to master...
